Create attributes `AsTwigFilter`, `AsTwigFunction` and `AsTwigTest` to ease extension development (#3916)
twigphp/Twig: doc/advanced.rst

and ``#[AsTwigTest]`` on any method of any class to define filters, functions, and tests.

Create a class, you don't need to extend any class or implement any interface
but it eases integration with frameworks if you use the attribute ``#[AsTwigExtension]``::

    use Twig\Attribute\AsTwigExtension;
    use Twig\Attribute\AsTwigFilter;
    use Twig\Attribute\AsTwigFunction;
    use Twig\Attribute\AsTwigTest;

    #[AsTwigExtension]
    class Project_Twig_Extension
    {
        #[AsTwigFilter('rot13')]
        public static function rot13(string $string): string
        {
            // ...
        }

        #[AsTwigFunction('lipsum')]
        public static function lipsum(int $count): string
        {
            // ...
        }

        #[AsTwigTest('even')]
        public static function isEven(int $number): bool
        {
            // ...
        }
    }

Then register the class using ``Twig\Extension\AttributeExtension``::

    $twig = new \Twig\Environment($loader);
    $twig->addExtension(new \Twig\Extension\AttributeExtension([
        Project_Twig_Extension::class,
    ]));

.. note::

>     The ``\Twig\Extension\AttributeExtension`` can be added only once to an environment.

Author: There is an exception if we try to register the extension a second time.

        // ...
    }

> Using PHP Attributes to define extensions

Reviewer: ```suggestion
Using PHP Attributes to define Extensions
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> .. versionadded:: 3.9

Reviewer: ```suggestion
.. versionadded:: 3.19
```

>     The attribute classes were added in Twig 3.9.

Reviewer: ```suggestion
    The attribute classes were added in Twig 3.19.
```

> From PHP 8.0, you can use the attributes ``#[AsTwigFilter]``, ``#[AsTwigFunction]``,

Reviewer: As we now require PHP 8, the version mention can be removed:

```suggestion
You can use the attributes ``#[AsTwigFilter]``, ``#[AsTwigFunction]``,
```

and ``#[AsTwigTest]`` on any method of any class to define filters, functions, and tests.

Create a class with the attribute ``#[AsTwigExtension]``::

    use Twig\Attribute\AsTwigExtension;
    use Twig\Attribute\AsTwigFilter;
    use Twig\Attribute\AsTwigFunction;
    use Twig\Attribute\AsTwigTest;

    #[AsTwigExtension]
    class ProjectExtension
    {
        #[AsTwigFilter('rot13')]
        public static function rot13(string $string): string
        {
            // ...
        }

        #[AsTwigFunction('lipsum')]
        public static function lipsum(int $count): string
        {
            // ...
        }

        #[AsTwigTest('even')]
        public static function isEven(int $number): bool
        {
            // ...
        }
    }

Then register the extension class::

    $twig = new \Twig\Environment($loader);
    $twig->addExtension(ProjectExtension::class);

> If all the methods are static, you are done. The ``Project_Twig_Extension`` class will

Reviewer: ```suggestion
If all the methods are static, you are done. The ``ProjectExtension`` class will
```

        // ...
    }

Using PHP Attributes to define Extensions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. versionadded:: 3.21

    The attribute classes were added in Twig 3.21.

You can use the attributes ``#[AsTwigFilter]``, ``#[AsTwigFunction]``, and
``#[AsTwigTest]`` on any method of any class to define filters, functions, and tests.

Create a class using this attributes::

    use Twig\Attribute\AsTwigFilter;
    use Twig\Attribute\AsTwigFunction;
    use Twig\Attribute\AsTwigTest;

    class ProjectExtension
    {
        #[AsTwigFilter('rot13')]
        public static function rot13(string $string): string
        {
            // ...
        }

        #[AsTwigFunction('lipsum')]
        public static function lipsum(int $count): string
        {
            // ...
        }

        #[AsTwigTest('even')]
        public static function isEven(int $number): bool
        {
            // ...
        }
    }

Then register the ``Twig\Extension\AttributeExtension`` with the class name::

    $twig = new \Twig\Environment($loader);

>     $twig->addExtension(new \Twig\Extension\AttributeExtension(ProjectExtension::class);

Reviewer: ```suggestion
    $twig->addExtension(new \Twig\Extension\AttributeExtension(ProjectExtension::class));
```

        // ...
    }

Using PHP Attributes to define Extensions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. versionadded:: 3.21

    The attribute classes were added in Twig 3.21.

You can use the attributes ``#[AsTwigFilter]``, ``#[AsTwigFunction]``, and

> ``#[AsTwigTest]`` on any method of any class to define filters, functions, and tests.

Reviewer 2: ```suggestion
You can add the ``#[AsTwigFilter]``, ``#[AsTwigFunction]``, and
``#[AsTwigTest]`` attributes to public methods of any class to define filters, functions, and tests.
```

> Create a class using this attributes::

Reviewer 2: ```suggestion
Create a class using these attributes::
```

    use Twig\Attribute\AsTwigFunction;

    class ProjectExtension
    {
        // Inject hypothetical dependencies
        public function __construct(private LipsumProvider $lipsumProvider) {}

        #[AsTwigFunction]
        public function lipsum(int $count): string
        {
            return $this->lipsumProvider->lipsum($count);
        }
    }

    $twig = new \Twig\Environment($loader);
    $twig->addExtension(new \Twig\Extension\AttributeExtension(ProjectExtension::class);
    $twig->addRuntimeLoader(new \Twig\RuntimeLoader\FactoryLoader([
        ProjectExtension::class => function () use ($lipsumProvider) {
            return new ProjectExtension($lipsumProvider);
        },
    ]));

The name can be customized to prevent conflict with other Twig callables and follow
the :doc:`Coding Standards<../coding_standards>`::

        class ProjectExtension
        {
            #[AsTwigFunction(name: 'name_in_snake_case')]
            public function nameInCamelCase(int $count): string
            {
                // ...
            }
        }

If you want to access the current environment instance in your filter or function,

> add the ``Twig\Environment`` type to the first argument of the method::

Reviewer 2: Does this automatically add the `needsEnvironment` option? What about `needsContext`? Is adding `array $context` enough?

Author: The detection of the context is less obvious (no specific type).
And it's very rarely used: `dump` is the only that uses it in all the Twig extra, Symfony and UX packages. Leaving it explicit seems to me to be more safe.

Reviewer 2: Agreed, just auto-detecting `needsEnvironment` seems best for these reasons.

Author: I did the detection initially: https://github.com/twigphp/Twig/blob/c5c00ed6c64ea2470f8a0dcb4c1fc18682f8fa5c/src/Extension/Extension.php

Reviewer: I'm confused. `needsEnvironment` is auto-detected, right?

Reviewer 2: `needsEnvironment` *is* auto-detected - `needsContext` is not.
